zonec.c

```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
+ */
+unsigned int
+zonec_read(
+	struct namedb *database,
+	struct domain_table *domains,
+	const char *name,
+	const char *zonefile,
+	struct zone *zone)
+{
+	const struct dname *origin;
+	zone_parser_t parser;
+	zone_options_t options = { 0 };
+	zone_name_buffer_t name_buffer;
+	zone_rdata_buffer_t rdata_buffer;
+	zone_buffers_t buffers = { 1, &name_buffer, &rdata_buffer };
+	struct zonec_state state = { database, domains, zone, NULL, 0, 0 };
+
+	origin = domain_dname(zone->apex);
+	options.origin.octets = dname_name(origin);
+	options.origin.length = origin->name_size;
+	options.default_ttl = DEFAULT_TTL;
+	options.default_class = CLASS_IN;
+	options.non_strict = zone_is_slave(zone->opts) != 0;
+	options.pretty_ttls = true; /* non-standard, for backwards compatibility */
+	options.log.categories = -1;
+	options.log.callback = &zonec_log;
+	options.accept.callback = &zonec_accept;
+
 	totalrrs = 0;
-	startzonec = time(NULL)+100000; /* disable */
-	parser_push_stringbuf(str);
-	yyparse();
-	parser_pop_stringbuf();
-	errors = parser->errors;
-	*num_rrs = totalrrs;
-	if(*num_rrs == 0)
-		*parsed = NULL;
-	else	*parsed = parser->prev_dname;
-	/* remove origin if it was not used during the parse */
-	if(parser->origin != error_domain)
-		domain_table_deldomain(parser->db, parser->origin);
-	region_free_all(parser->rr_region);
-	zonec_desetup_string_parser();
-	parser_flush();
-	return errors;
-}
+	startzonec = time(NULL);
 
-/** check SSHFP type for failures and emit warnings */
-void check_sshfp(void)
-{
-	uint8_t hash;
-	uint16_t size;
-	if(parser->current_rr.rdata_count < 3)
-		return; /* cannot check it, too few rdata elements */
-	if(!parser->current_rr.rdatas[0].data ||
-		!parser->current_rr.rdatas[1].data ||
-		!parser->current_rr.rdatas[2].data ||
-		!parser->current_rr.owner)
-		return; /* cannot check, NULLs (due to earlier errors) */
-	if(rdata_atom_size(parser->current_rr.rdatas[1]) != 1)
-		return; /* wrong size of the hash type rdata element */
-	hash = rdata_atom_data(parser->current_rr.rdatas[1])[0];
-	size = rdata_atom_size(parser->current_rr.rdatas[2]);
-	if(hash == 1 && size != 20) {
-		zc_warning_prev_line("SSHFP %s of type SHA1 has hash of "
-			"wrong length, %d bytes, should be 20",
-			domain_to_string(parser->current_rr.owner),
-			(int)size);
-	} else if(hash == 2 && size != 32) {
-		zc_warning_prev_line("SSHFP %s of type SHA256 has hash of "
-			"wrong length, %d bytes, should be 32",
-			domain_to_string(parser->current_rr.owner),
-			(int)size);
+	/* Parse and process all RRs.  */
+	if (zone_parse(&parser, &options, &buffers, zonefile, &state) != 0) {
+		return state.errors;
+	}
+
+	/* Check if zone file contained a correct SOA record */
+	if (!zone) {
+		log_msg(LOG_ERR, "zone configured as '%s' has no content.", name);
+		state.errors++;
+	} else if (!zone->soa_rrset || zone->soa_rrset->rr_count == 0) {
+		log_msg(LOG_ERR, "zone configured as '%s' has no SOA record", name);
+		state.errors++;
+#if 0
```

Comment:
This check is not disabled in the original zonec_read function.

Reply:
Correct. I disabled it for now. I'll have to turn the owner into a `dname` and then compare. I'll fix in an upcoming commit.

Reply:
I've re-enabled this check btw. Comparing the owner to the origin, which is constructed from the zone name.